### PR TITLE
De-dupe gradient stops

### PR DESCRIPTION
**Pull request by gpeal**
Refer to code comments for more info:
Before:
CleanShot 2022-05-26 at 11 25 30
After:
CleanShot 2022-05-26 at 11 26 54

### CODE REVIEW

File: lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java [hunk 1]
```diff
@@ -223,4 +221,60 @@ private int getColorInBetweenOpacityStops(float position, int color, float[] opa
...
+    }
+
+    if (numDuplicates == 0) {
+      return mergedNotTruncated;
+    }
+
+
+    float[] merged = new float[mergedNotTruncated.length - numDuplicates];
+    System.arraycopy(mergedNotTruncated, 0, merged, 0, merged.length);
+    return merged;
```
Comment (denis-bezrukov) [inline lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java:278]: nit: 
```suggestion
    return Arrays.copyOf(mergedNotTruncated, mergedNotTruncated.lengh - numDuplicates);
```

File: lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java [hunk 2]
```diff
@@ -223,4 +221,60 @@ private int getColorInBetweenOpacityStops(float position, int color, float[] opa
...
+
+      final float b;
+      if (bIndex < arrayB.length) {
+        b = arrayB[bIndex];
+      } else {
+        b = Float.NaN;
+      }
+
+      if (Float.isNaN(b) || a < b) {
+        mergedNotTruncated[mergedIndex++] = a;
```
Comment (denis-bezrukov) [inline lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java:258]: isn't `mergeIndex` can be replaced by `i`?

File: lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java [hunk 3]
```diff
@@ -223,4 +221,60 @@ private int getColorInBetweenOpacityStops(float position, int color, float[] opa
...
+    // This will be the merged list but may be longer than what is needed if there are duplicates.
+    // If there are, the 0 elements at the end need to be truncated.
+    float[] mergedNotTruncated = new float[arrayA.length + arrayB.length];
+    for (int i = 0; i < mergedNotTruncated.length; i++) {
+      final float a;
+      if (aIndex < arrayA.length) {
+        a = arrayA[aIndex];
+      } else {
+        a = Float.NaN;
+      }
```
Comment (denis-bezrukov) [inline lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java:248]: nit could be elvis

```suggestion
      final float a = aIndex < arrayA.length ? arrayA[aIndex] : Float.NaN
```

**Review by denis-bezrukov — APPROVED** [review]

File: lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java (hunk 1 above)
Comment (gpeal) [reply lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java:278]: Done!

File: lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java (hunk 2 above)
Comment (gpeal) [reply lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java:258]: Yup!

File: lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java (hunk 3 above)
Comment (gpeal) [reply lottie/src/main/java/com/airbnb/lottie/parser/GradientColorParser.java:248]: 👍

**Comment by Vodanhlatui** [comment]
> Tham khảo các nhận xét về mã để biết thêm thông tin: Trước: Sau: CleanShot 2022-05-26 lúc 11 25 30 CleanShot 2022-05-26 lúc 11 26 54

**Comment by imbryk** [comment]
@gpeal @denis-bezrukov 
This PR is crashing the app for quite a few complex animations we use.
I'm not able to identify the place in our animations, as they are really big, but I've found that the problem is that this PR  changes the length of color array in `GradientColor` which results in https://github.com/airbnb/lottie-android/blob/master/lottie/src/main/java/com/airbnb/lottie/animation/keyframe/GradientColorKeyframeAnimation.java being created with keyframes which have GradientColors of different lengths, which eventually crashes at the `GradientColor.lerp` method with `java.lang.ArrayIndexOutOfBoundsException`

This can theoretically be simply fixed not to crash, but I believe that would completely break the intended gradient animation.

If the problem solved by this PR was needed mostly for static lottie files, maybe you could apply this change only when parsing a static value:
https://github.com/airbnb/lottie-android/blob/master/lottie/src/main/java/com/airbnb/lottie/parser/KeyframeParser.java#L369
but when parsing animation keep the old way of parsing it (i.e. keeping all values)

**Comment by imbryk** [comment]
@gpeal would it be possible to send it to you privately?

**Comment by Blincwear88** [comment]
#2297

**Comment by gpeal** [comment]
@imbryk Yeah, go ahead and DM it to me on Twitter but open a new issue saying you did that so it doesn't slip through the cracks the next time I work on this.

**Comment by imbryk** [comment]
@gpeal I am not able to send you a message on twitter 🤷 - please grab the files from gdrive:
https://drive.google.com/drive/folders/1LOiSaIJq1LJL848Ne8rA9Omrn2ErynfY

**Pull request merged**